src/app/archive/lib/types.ml

```diff
@@ -0,0 +1,250 @@
+open Core
+open Coda_base
+open Signature_lib
+
+(** Library used to encode and decode types to a graphql format. 
+    
+    Unfortunately, the graphql_ppx does not have an "encode" attribute that allows us to map an OCaml type into some graphql schema type in a clean way. Therefore, we have to make our own encode and decode functions. On top of this, the generated GraphQL schema constructed by Hasura creates insert methods where the input fields are optional, even though the inputs are explicitly labeled as NOT NULL in Postgres. Additionally, graphql_ppx does not let us redefine a new type to make these optional inputs mandatory. Therefore, we are forced to lift the option types to some of our types. Furthermore, some types that are in Postgres but are not GraphQL primitive types are treated as custom scalars. These types include `bigint`, `bit(n)` and enums. graphql_ppx treats custom scalars as Yojson.Basic.t types (usually they are encoded as Json string types). 
+    
+    As a result, encoding a type to a Hasura input is broken into two phases: 
+    1. Postgres phase: This phase converts OCaml type (such as blocks and transactions) to some intermediate representation that is very similar to their respective Postgres schema type. 
+    2. Graphql object phase: The intermediate Postgres types are converted directly to some input that the Hasura Graphql schema would accept. This is essentially lifting types with the Option type and coercing Postgres custom scalar into Yojson types 
```

Comment:
I'm a little confused by this as it seems like as far as the client is concerned, it's just talking to a graphql endpoint so not sure what's changed by there being postgres underneath.

Hard to follow everything going on in this file, but fwiw I think we've had a decent amount of success in the other graphql code by having simple modules of of `Encoders` and `Decoders`.

Reply:
Second -- @wu-s-john does anything consume the "postgres value" explicitly? Maybe we can remove the intermediate representation

Reply:
I think having the `encode` and `decode` functions for a specific type grouped together is actually quite nice here as they're more complex than where we used `Encoders` and `Decoders` in the reason code (disagree with that one @Schmavery ).

I didn't have too hard of a time following -- esp. if we replace the bitstring stuff

Reply:
especially the user_command here, but maybe some of the other ones could be similar to that:
see https://github.com/CodaProtocol/coda/pull/3447/files/485d6f21855b667a51c5bbabb2d1f77fc4b159c2#diff-fd9937bf15e6215b5186c5bdb4c0a887R68